### PR TITLE
feat: (experimental) show pinning remediation advice for Python

### PR DESCRIPTION
**Pull request by kyegupov**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
Shows Python remediation advice in `snyk test`, based on pinning.
Pinning means directly specifying the needed version, even for transitive dependencies.

This depends on the Registry changes: https://github.com/snyk/registry/pull/9434
The advice is calculated and shown differently, based on whether the `actionableCliRemediation` flag is set.

#### Screenshots:

With actionableCliRemediation:
image

Without actionableCliRemediation:
image

### CODE REVIEW

File: src/lib/feature-flags.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import request = require('./request');
+import snyk = require('.'); // TODO(kyegupov): fix import
```
Comment (dkontorovskyy) [inline src/lib/feature-flags.ts]: maybe time had come ? :)

Comment (kyegupov) [reply src/lib/feature-flags.ts]: no :( that index.js file is absolutely insane.

File: src/lib/snyk-test/run-test.ts [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import * as _ from 'lodash';
-import fs = require('then-fs');
+import * as fs from 'fs';
```
Comment (dkontorovskyy) [inline src/lib/snyk-test/run-test.ts]: 😍

File: src/cli/commands/test/formatters/legacy-format-issue.ts [hunk 3]
```diff
@@ -135,12 +139,25 @@ function createFixedInText(vuln: any): string {
...
+  packageManager: SupportedPackageManagers,
+  pinningSupported: boolean,
+): string {
   let wizardHintText = '';
   if (WIZARD_SUPPORTED_PACKAGE_MANAGERS.includes(packageManager)) {
     wizardHintText = 'Run `snyk wizard` to explore remediation options.';
   }
-
+  if (pinningSupported
+    && (vuln.nearestFixedInVersion || vuln.fixedIn)
```
Comment (lili2311) [inline src/cli/commands/test/formatters/legacy-format-issue.ts]: `nearestFixedInVersion` is docker specific I believe

**Comment by lili2311** [comment]
screenshots please

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 4]
```diff
@@ -33,13 +45,37 @@ export function formatIssuesWithRemediation(
       name: vuln.name,
       version: vuln.version,
       fixedIn: vuln.fixedIn,
+      paths: vuln.list.map((v) => v.from),
     };
   }
   const results = [chalk.bold.white('Remediation advice')];
 
-  const upgradeTextArray = constructUpgradesText(remediationInfo.upgrade, basicVulnInfo);
```
Comment (lili2311) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: I feel like pinning can be it's own function that generates pinning? `constructPinOrUpgradesText` why not have construct pinning advice? 

Is pinning not it's own section that is for packages that have no upgrade paths? It doesn't seem right to prefer pinning to upgrading? Is there some stuff to read on this feature somewhere to gain more context and what Python remediation will do?

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 5]
```diff
@@ -97,32 +133,116 @@ function constructUpgradesText(
...
+        basicVulnInfo[id].severity,
+        basicVulnInfo[id].isNew,
+        `${basicVulnInfo[id].name}@${basicVulnInfo[id].version}`))
       .join('\n');
     upgradeTextArray.push(upgradeText + thisUpgradeFixes);
   }
   return upgradeTextArray;
 }
 
+function constructPinOrUpgradesText(
```
Comment (lili2311) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: this is huge! Can it not be split up for Pin vs Upgrade, I feel like they are completely different?

Comment (kyegupov) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: Was split, common formatting code was extracted.

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts (hunk 4 above)
Comment (kyegupov) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: With Python, we don't have full ecosystem information, but we can perform pinning.
Thus, we chose to do pinning _insteat_ of upgrade paths.
So, it's either-or.

File: src/cli/commands/test/formatters/legacy-format-issue.ts (hunk 3 above)
Comment (kyegupov) [reply src/cli/commands/test/formatters/legacy-format-issue.ts]: Extracted the field into a Docker-specific type.

Comment (lili2311) [reply src/cli/commands/test/formatters/legacy-format-issue.ts]: I mean surely we should not be checking it here: `if (pinningSupported
    && vuln.fixedIn)` is all we need.

Comment (kyegupov) [reply src/cli/commands/test/formatters/legacy-format-issue.ts]: yep, also done that :)

File: src/cli/commands/test/formatters/legacy-format-issue.ts [hunk 6]
```diff
@@ -157,7 +170,7 @@ function createRemediationText(vuln, packageManager) {
         const selfUpgradeInfo = (v.upgradePath.length > 0)
           ? ` (triggers upgrades to ${ v.upgradePath.join(' > ')})`
           : '';
-        const testedPackageName = v.upgradePath[0].split('@');
+        const testedPackageName = parsePackageNameVersion(v.upgradePath[0] as string);
```
Comment (lili2311) [inline src/cli/commands/test/formatters/legacy-format-issue.ts]: why the change?

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 7]
```diff
@@ -36,13 +50,37 @@ export function formatIssuesWithRemediation(
...
+      paths: vuln.list.map((v) => v.from),
     };
   }
   const results = [chalk.bold.white('Remediation advice')];
 
-  const upgradeTextArray = constructUpgradesText(remediationInfo.upgrade, basicVulnInfo);
-  if (upgradeTextArray.length > 0) {
-    results.push(upgradeTextArray.join('\n'));
+  if (remediationInfo.pin && Object.keys(remediationInfo.pin).length) {
+    const upgradesByCulprit: UpgradesByCulprit = {};
```
Comment (lili2311) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: `upgradesByCulprit` is there a nicer name for this? What is it trying to name? Is it `transitive` vs `direct`?

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 8]
```diff
@@ -36,13 +50,37 @@ export function formatIssuesWithRemediation(
...
+    for (const topLvlPkg of Object.keys(remediationInfo.upgrade)) {
```
Comment (lili2311) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: `topLvlPkg` `topLevelPackage` is nicer, let's avoid shortening as it is harder to read

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts [hunk 9]
```diff
@@ -36,13 +50,37 @@ export function formatIssuesWithRemediation(
...
+          upgradesByCulprit[targetPkgStr] = [];
+        }
+        upgradesByCulprit[targetPkgStr].push({
+          name: topLvlPkg,
+          version: remediationInfo.upgrade[topLvlPkg].upgradeTo,
+        });
+      }
+    }
+    const upgradeTextArray = constructPinText(remediationInfo.pin, upgradesByCulprit, basicVulnInfo);
+    if (upgradeTextArray.length > 0) {
```
Comment (lili2311) [inline src/cli/commands/test/formatters/remediation-based-format-issues.ts]: this is repeated twice, can just do it once after this whole if?

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts (hunk 7 above)
Comment (kyegupov) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: Ok, will rename to "byAffectedPackage".

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts (hunk 8 above)
Comment (kyegupov) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: decided on topLevelPkg. First, saves some characters, second, since `package` is a reserved word, shortening to `pkg` is a good habit anyway.

File: src/cli/commands/test/formatters/remediation-based-format-issues.ts (hunk 9 above)
Comment (kyegupov) [reply src/cli/commands/test/formatters/remediation-based-format-issues.ts]: Done.
